### PR TITLE
Reconnect gRPC on error

### PR DESCRIPTION
**Pull request by mpeterss**

### CODE REVIEW

**Comment by mpeterss** [comment]
https://github.com/ostelco/ostelco-core/issues/102

**Comment by mpeterss** [comment]
>Should we call reconnect on any specific type of Throwable/Exception, or should it be called for any type?

I only see "io.grpc.StatusRuntimeException: UNAVAILABLE: io exception", but maybe I should start with only that one then

> If the onError event occurs how do we guarantee that messages are not lost, and they are sent again after reconnection?

Currently all requests are lost when gRPC is down, nothing is buffered. If we buffer, we need to figure out how we long should store a request before we throw it away.

File: ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java [hunk 1]
```diff
@@ -77,9 +81,22 @@ protected boolean removeEldestEntry(Map.Entry<String, SessionContext> eldest) {
...
-    private abstract class AbstactObserver<T> implements StreamObserver<T> {
+    private abstract class CreditControlRequestObserver<T> implements StreamObserver<T> {
+        public final void onError(Throwable t) {
+            LOG.error("We got an error", t);
+            reconnectCreditControlRequest();
+        }
+
+        public final void onCompleted() {
+            // Nothing to do here
+            LOG.info("It seems to be completed");
```
Comment (prasanthu) [inline ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java]: Looks unnecessary

File: ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java [hunk 2]
```diff
@@ -77,9 +81,22 @@ protected boolean removeEldestEntry(Map.Entry<String, SessionContext> eldest) {
     };
 
 
-    private abstract class AbstactObserver<T> implements StreamObserver<T> {
+    private abstract class CreditControlRequestObserver<T> implements StreamObserver<T> {
+        public final void onError(Throwable t) {
+            LOG.error("We got an error", t);
```
Comment (prasanthu) [inline ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java]: Can you change the text to anything other than `We got an error` ;-)

File: ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java (hunk 1 above)
Comment (mpeterss) [reply ocsgw/src/main/java/org/ostelco/ocsgw/data/grpc/GrpcDataSource.java]: Yes I just wanted to check if it ever got called. So far it has never been. I will remove it

**Pull request merged**